src/ape_cache/query.py

```diff
@@ -0,0 +1,102 @@
+from functools import partial
+from pathlib import Path
+from typing import Any, Dict, Optional
+
+import pandas as pd
+from pydantic import BaseModel
+from sqlalchemy import create_engine
+
+from ape.api import QueryAPI, QueryType
+from ape.api.query import AccountQuery, BlockQuery, ContractEventQuery, _BaseQuery
+from ape.exceptions import QueryEngineError
+from ape.utils import cached_property, singledispatchmethod
+
+TABLE_NAME = {
+    BlockQuery: "blocks",
+    AccountQuery: "transactions",
+    ContractEventQuery: "contract_events",
+}
+
+
+def get_columns_from_item(query: _BaseQuery, item: BaseModel) -> Dict[str, Any]:
+    return {k: v for k, v in item.dict().items() if k in query.columns}
+
+
+class CacheQueryProvider(QueryAPI):
+    """
+    Default implementation of the ape.api.query.QueryAPI
+    Allows for the query of blockchain data using connected provider
+    """
+
+    @property
+    def database_file(self) -> Path:
+        return self.config_manager.DATA_FOLDER / "cache.db"
+
+    @cached_property
+    def engine(self):
+        return create_engine(f"sqlite:///{self.database_file}", pool_pre_ping=True)
+
+    @singledispatchmethod
+    def estimate_query(self, query: QueryType) -> Optional[int]:
+        return None  # can't handle this query
+
+    @estimate_query.register
+    def estimate_block_query(self, query: BlockQuery) -> Optional[int]:
+        with self.engine.connect() as conn:
+            q = conn.execute(
+                f"""
+                SELECT COUNT(*) 
+                FROM blocks 
+                WHERE blocks.number >= {query.start_block} 
```

Comment:
You should generally not use string interpolation in SQL queries to prevent injecting bad stuff.

I think you can rewrite this like this:

```python
conn.execute(
                """
                SELECT COUNT(*) 
                FROM blocks 
                WHERE blocks.number >= %(start_block)s
                AND blocks.number <= %(stop_block)s
                """, {start_block=query.start_block, stop_block=query.stop_block}
)
```

^ better practice IICR and more security-minded

Reply:
The query calls are coming from within the framework, so I don't necessarily think this is a problem. There should be no foreign injection of `QueryType` objects using this plugin. The relative performance also suffers from doing it this way:
https://twitter.com/raymondh/status/1205969258800275456